spring-kafka/src/main/java/org/springframework/kafka/ReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,304 @@
+	public RequestReplyFuture<K, V, R> sendAndReceive(ProducerRecord<K, V> record) {
+		CorrelationKey correlationId = createCorrelationId();
+		record.headers().add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.correlationId));
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("Sending: " + record + " with correlationId: " + correlationId);
+		}
+		TemplateRequestReplyFuture<K, V, R> future = new TemplateRequestReplyFuture<>();
+		this.futures.put(correlationId, future);
+		future.setSendFuture(send(record));
+		this.scheduler.schedule(() -> {
+			RequestReplyFuture<K, V, R> removed = this.futures.remove(correlationId);
+			if (removed != null) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("Reply timed out for: " + record + " with correlationId: " + correlationId);
+				}
+				removed.setException(new KafkaException("Reply timed out"));
+			}
+		}, Instant.now().plusMillis(this.replyTimeout));
+		return future;
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		if (!this.schedulerSet) {
+			((ThreadPoolTaskScheduler) this.scheduler).destroy();
+		}
+	}
+
+	/**
+	 * Subclasses can override this to generate custom correlation ids.
+	 * The default implementation is a 16 byte representation of a UUID.
+	 * @return the key.
+	 */
+	protected CorrelationKey createCorrelationId() {
+		UUID uuid = UUID.randomUUID();
+		byte[] bytes = new byte[16];
+		ByteBuffer bb = ByteBuffer.wrap(bytes);
+		bb.putLong(uuid.getMostSignificantBits());
+		bb.putLong(uuid.getLeastSignificantBits());
+		return new CorrelationKey(bytes);
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, R>> data) {
+		data.forEach(record -> {
+			Iterator<Header> iterator = record.headers().iterator();
+			CorrelationKey correlationId = null;
+			while (correlationId == null && iterator.hasNext()) {
+				Header next = iterator.next();
+				if (next.key().equals(KafkaHeaders.CORRELATION_ID)) {
+					correlationId = new CorrelationKey(next.value());
+				}
+			}
+			if (correlationId == null) {
+				this.logger.error("No correlationId found in reply: " + record);
+			}
+			else {
+				RequestReplyFuture<K, V, R> future = this.futures.remove(correlationId);
+				if (future == null) {
+					this.logger.error("No pending reply: " + record + " with correlationId: "
+							+ correlationId + ", perhaps timed out");
+				}
+				else {
+					if (this.logger.isDebugEnabled()) {
+						this.logger.debug("Received: " + record + " with correlationId: " + correlationId);
+					}
+					future.set(record);
+				}
+			}
+		});
+	}
+
+	/**
+	 * Wrapper for byte[] that can be used as a hash key. We could have used BigInteger
+	 * instead but this wrapper is less expensive. We do use a BigInteger in
+	 * {@link #toString()} though.
+	 */
+	public static final class CorrelationKey {
+
+		private final byte[] correlationId;
+
+		private volatile Integer hashCode;
+
+		public CorrelationKey(byte[] bytes) {
+			this.correlationId = bytes;
```

Comment:
`Assert.notNull()` - this is public API, so end-user may decide to try to suffer of `NPE` 😄